### PR TITLE
[github-scanner] introduce new integration

### PR DESCRIPTION
**Pull request by maorfr**
This integration scans github repos (we can change it later to include gitlab) for leaked credentials and creates PRs in app-interface to remove leaked creds if they are found.

it uses [git-secrets](https://github.com/awslabs/git-secrets) under the hood, and we can add any other tool that we want in addition.

### CODE REVIEW

File: utils/git_secrets.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    logging.info('found suspects in {}'.format(repo_url))
+    suspcted_files = get_suspected_files(err)
+    leaked_keys = get_leaked_keys(repo, suspcted_files, existing_keys)
+    if leaked_keys:
+        logging.info('found suspected leaked keys: {}'.format(leaked_keys))
+
+    return leaked_keys
+
+
+def get_suspected_files(error):
```
Comment (jmelis) [inline utils/git_secrets.py]: can you drop an example of the output so we can understand that parsing?

File: utils/git_secrets.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        if e.startswith('warning'):
+            continue
+        commit_path_split = e.split(' ')[0].split(':')
+        commit, path = commit_path_split[0], commit_path_split[1]
+
+        suspects.append((commit, path))
+    return set(suspects)
+
+
+def get_leaked_keys(repo, suspcted_files, existing_keys):
```
Comment (jmelis) [inline utils/git_secrets.py]: typo in both places, no?

File: setup.py [hunk 3]
```diff
@@ -26,6 +26,7 @@
         "ldap3>=2.5.2,<2.6.0",
         "anymarkup>=0.7.0,<0.8.0",
         "python-gitlab>=1.11.0,<1.12.0",
+        "GitPython>=2.1.14,<3.0.0",
```
Comment (jmelis) [inline setup.py]: I'm not sure we need this, you are using git in two ways: 1) to git clone and 2) to get a file from a previous commit. These two things can be easily done with subprocess, no?

for reference:
```
git show <commit>:<path>
```

File: reconcile/github_scanner.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    all_leaked_keys = [key for keys in results for key in keys]
+
+    deleted_keys = aws_sos.get_deleted_keys(accounts)
+    keys_to_delete = \
+        [{'account': account, 'key': key}
+         for key in all_leaked_keys
+         for account, user_keys in existing_keys.items()
+         if key in [uk for uks in user_keys.values() for uk in uks]
+         and key not in deleted_keys[account]]
+    aws_sos.act(dry_run, gitlab_project_id, gqlapi, accounts, keys_to_delete)
```
Comment (jmelis) [inline reconcile/github_scanner.py]: nice refactor :)

File: reconcile/github_scanner.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+def run(gitlab_project_id, dry_run=False, thread_pool_size=10):
+    gqlapi = gql.get_api()
+    accounts = gqlapi.query(AWS_ACCOUNTS_QUERY)['accounts']
+    aws = AWSApi(thread_pool_size, accounts)
+    existing_keys = aws.get_users_keys()
+    existing_keys_list = [key for user_key in existing_keys.values()
+                          for keys in user_key.values() for key in keys]
+
+    app_int_github_repos = \
+        gitlab_permissions.get_repos(gqlapi, server='https://github.com')
```
Comment (jmelis) [inline reconcile/github_scanner.py]: a bit ugly, can we refactor?

File: utils/git_secrets.py (hunk 1 above)
Comment (maorfr) [reply utils/git_secrets.py]:
```
$ git-secrets --scan-history
9e43530e337d7d0feec726db13d422a78939b183:training/datastore/get_preprocess_data.py:33:                 aws_secret_access_key='2+Po/7pyFzSNnHFfAQDbLpZ6p/0uHE5Ujx7mE1W7',

[ERROR] Matched one or more prohibited patterns

Possible mitigations:
- Mark false positives as allowed using: git config --add secrets.allowed ...
- Mark false positives as allowed by adding regular expressions to .gitallowed at repository's root directory
- List your configured patterns: git config --get-all secrets.patterns
- List your configured allowed patterns: git config --get-all secrets.allowed
- List your configured allowed patterns in .gitallowed at repository's root directory
- Use --no-verify if this is a one-time false positive
```

File: reconcile/github_scanner.py (hunk 4 above)
Comment (maorfr) [reply reconcile/github_scanner.py]: thanks!

File: utils/git_secrets.py (hunk 2 above)
Comment (maorfr) [reply utils/git_secrets.py]: thanks!

File: reconcile/github_scanner.py (hunk 5 above)
Comment (maorfr) [reply reconcile/github_scanner.py]: handling this in a separate PR!

**Comment by maorfr** [comment]
comments addressed in #260

**Pull request merged**